src/DotNetty.Common/Concurrency/SingleThreadEventExecutor.cs

```diff
@@ -356,33 +359,32 @@ bool RunAllTasks(PreciseTimeSpan timeout)
             return true;
         }
 
-        void FetchFromScheduledTaskQueue()
+        bool FetchFromScheduledTaskQueue()
         {
-            if (this.HasScheduledTasks())
+            PreciseTimeSpan nanoTime = PreciseTimeSpan.FromStart;
+            IScheduledRunnable scheduledTask = this.PollScheduledTask(nanoTime);
+            while (scheduledTask != null)
             {
-                PreciseTimeSpan nanoTime = PreciseTimeSpan.FromStart;
-                while (true)
+                if (!this.taskQueue.TryEnqueue(scheduledTask))
                 {
-                    IScheduledRunnable scheduledTask = this.PollScheduledTask(nanoTime);
-                    if (scheduledTask == null)
-                    {
-                        break;
-                    }
-
-                    this.taskQueue.TryEnqueue(scheduledTask);
+                    // No space left in the task queue add it back to the scheduledTaskQueue so we pick it up again.
+                    this.ScheduledTaskQueue.Enqueue(scheduledTask);
```

Comment:
`scheduledTaskQueue` is a priority queue. While reordering is possible amid items with the same time (down to ticks), users should not rely on this queue for ordering really.